feat: issue 3889 signing stacks transactions rpc method

fess-v opened the pull request:
Hello! This is a PR in relation to this [issue](https://github.com/hirosystems/wallet/issues/3889). It's the first draft so many things can change.
I significantly changed the spec which we discussed in the issue since I haven't considered a lot of things, therefore attaching the new spec:
```javascript
const rpcSignStacksTransactionParamsSchema = yup.object().shape({
  stxAddress: yup.string().required(),
  serializedTransactionHex: yup.string().required(),
  disableFeeSelection: yup.boolean(),
  disableNonceSelection: yup.boolean(),
  attachment: yup.string(),
  appDetails: yup.object({
    name: yup.string(),
    icon: yup.string()
  }),
  network: yup.string().oneOf(StacksNetworks), // mainnet | testnet | devnet | mocknet
});
```
new fields:
1. `disableFeeSelection` and `disableNonceSelection` - required for multisig transactions since owners of a multisig except the first one cannot modify these parameters and shouldn't be able to do it
2. `attachment` - an already existing field in the previous transaction signing approach, does the same thing
3. `appDetails` - same as attachment, was already present before

Some points to discuss on this PR:
1. 
With disabled fee selection looks like that:
![image](https://github.com/hirosystems/wallet/assets/26626714/03a0834e-0dcb-4e91-bd04-fd7db946bc1a)
It may need some styling work.
2. Changed the structure, moved most components for the previous signing window from `pages` into `features` since more than one page now requires these components
3. Used `cleanMemoString` to remove redundant padding added when serializing a TX using the standard way from Stacks.js
![image](https://github.com/hirosystems/wallet/assets/26626714/aeccd687-eb66-44d4-9269-f881de565c39)
It's done in this method, so to display the memo string as it was initially we need to clean this field
4, Added `stacksTransactionContext` for making a `transactionRequest` variable available for viewing in all components, made it because if we put the same request object to URL, we will get two sources of truth in the URL - one is serialized transaction to sign, the second one is this request. So instead I put to this new context where I need to define this variable for viewing
5. Let's also discuss the further improvement of Stacks.js and micro-stacks packages, so we can add a similar universal method for signing multisig and singlesig transactions and parameters like disabling fee and nonce selection.

Tested this PR on Asigna - multisig transactions worked fine as well as singlesig transactions but anyway requires additional testing.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/hirosystems/wallet?pullRequest=4048) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/hirosystems/wallet?pullRequest=4048) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/hirosystems/wallet?pullRequest=4048) it.</sub>

kyranjamie commented on src/app/components/nonce-setter.tsx:
```diff
@@ -8,18 +8,21 @@ import { useNextNonce } from '@app/query/stacks/nonce/account-nonces.hooks';
+  skipNonceSetting?: boolean;
 }
-export function NonceSetter({ children }: NonceSetterProps) {
+export function NonceSetter({ children, skipNonceSetting }: NonceSetterProps) {
```
Perhaps we could just conditionally not render this component? 

`!skipNonceSetting && <NonceSetter />`?

kyranjamie commented on src/app/pages/rpc-sign-stacks-transaction/rpc-sign-stacks-transaction.tsx:
```diff
@@ -0,0 +1,176 @@
+  if (!transactionRequest) return null;
+
+  return (
+    <StacksTransactionProvider value={transactionRequest}>
+      <StacksTransactionSigner onSignStacksTransaction={onSignStacksTransaction}
+                               onCancel={onCancel}
+                               stacksTransaction={stacksTransaction}
+                               disableFeeSelection={disableFeeSelection}
+                               disableNonceSelection={disableNonceSelection}/>
+    </StacksTransactionProvider>
+  );
```
You'll need to run prettier to follow our formatting

kyranjamie commented on src/app/pages/rpc-sign-stacks-transaction/rpc-sign-stacks-transaction.tsx:
```diff
@@ -0,0 +1,176 @@
+  );
+}
+
+function useRpcSignPsbt() {
```
copy paste, to rename

kyranjamie commented on src/app/components/fees-row/fees-row.tsx:
```diff
@@ -34,6 +37,12 @@ export function FeesRow(props: FeeRowProps): React.JSX.Element {
   const isCustom = feeTypeField.value === FeeTypes[FeeTypes.Custom];
   const selectedItem = Number(FeeTypes[feeTypeField.value]);
 
+  useEffect(() => {
+    if (props.defaultFeeValue) {
+      setTimeout(() => handleSelectFeeEstimateOrCustomField(FeeTypes.Custom));
+    }
+  }, [props.defaultFeeValue])
```
This data flow of having an effect inside the FeeRow component set a fee, even though it's disabled, is hard to follow. 

Any other ways we can do this?

kyranjamie reviewed (commented):
Are these two properties really things we need to leave to the developer to set themselves

```
  disableFeeSelection: yup.boolean(),
  disableNonceSelection: yup.boolean(),
```

Could this be done another way, such as detecting multisig?

fess-v commented:
> Are these two properties really things we need to leave to the developer to set themselves
> 
> ```
>   disableFeeSelection: yup.boolean(),
>   disableNonceSelection: yup.boolean(),
> ```
> 
> Could this be done another way, such as detecting multisig?

yes, I think it is better to detect if this transaction is a multisig AND was already signed by someone, in this case it won't be possible to modify nonce and fees
thanks!

kyranjamie commented on src/app/components/fees-row/fees-row.tsx:
```diff
@@ -34,6 +37,8 @@ export function FeesRow(props: FeeRowProps): React.JSX.Element {
   const isCustom = feeTypeField.value === FeeTypes[FeeTypes.Custom];
   const selectedItem = Number(FeeTypes[feeTypeField.value]);
 
+  useEffect(() => {}, [props.defaultFeeValue]);
```
Does nothing, to remove

kyranjamie commented on src/shared/rpc/methods/sign-stacks-transaction.ts:
```diff
@@ -0,0 +1,40 @@
+  stxAddress: yup.string().required(),
+  serializedTransactionHex: yup.string().required(),
+  attachment: yup.string(),
+  appDetails: yup.object({
+    name: yup.string(),
+    icon: yup.string(),
+  }),
```
`appDetails` is an old Stacks request structure. Maybe we can just pass an `icon`?

Imo we should drop `name`. It's dangerous as it can be used to trick users. Honestly, even can the icon. Right now we use the favicon of the `origin`, so maybe we can remove entirely?

kyranjamie commented on src/shared/rpc/methods/sign-stacks-transaction.ts:
```diff
@@ -0,0 +1,40 @@
+
+const rpcSignStacksTransactionParamsSchema = yup.object().shape({
+  stxAddress: yup.string().required(),
+  serializedTransactionHex: yup.string().required(),
```
I wonder if just `transactionHex` is okay? Are there other ways to format a stacks transaction in hex other than to serialise it? If not, it's not necessary to specify that it must be serialised?

kyranjamie commented on src/background/messaging/rpc-methods/sign-stacks-transaction.ts:
```diff
@@ -0,0 +1,106 @@
+  triggerRequestWindowOpen,
+} from '../messaging-utils';
+
+function validateStacksTransaction(serializedTransactionHex: string) {
+  try {
+    deserializeTransaction(serializedTransactionHex);
+    return true;
+  } catch (e) {
+    return false;
+  }
+}
```
isn't this exported from stacks.js? 🤔

kyranjamie reviewed (commented):
A few more comments. Would be great to get @fbwoolf to review this as well, as she's familiar with much of the custom fee work for stacks transactions

kyranjamie commented:
Looks pretty good @fess-v, thanks. Would you also be able to draft up some documentation we'd be able to use on our docs site? https://hirowallet.gitbook.io/developers/

kyranjamie commented on src/app/features/stacks-transaction-request/components/stacks-transaction.context.ts:
```diff
@@ -0,0 +1,7 @@
+
+import { TransactionPayload } from '@stacks/connect';
+
+export const stacksTransactionContext = createContext<TransactionPayload>(undefined as any);
```
```suggestion
export const stacksTransactionContext = createContext<TransactionPayload | null>(null);
```

kyranjamie commented on src/app/pages/rpc-sign-stacks-transaction/rpc-sign-stacks-transaction.tsx:
```diff
@@ -0,0 +1,185 @@
+        makeRpcSuccessResponse('signStacksTransaction', {
+          id: requestId,
+          result: {
+            transactionHex: Buffer.from(signedTransaction.serialize()).toString('hex'),
```
Better to favour uin8array so

```ts
bytesToHex(signedTransaction.serialize())
```

fbwoolf commented on src/app/components/fees-row/components/fee-estimate-item.tsx:
```diff
@@ -30,13 +31,13 @@ export function FeeEstimateItem(props: FeeEstimateItemProps) {
       mb="0px !important"
       minWidth="100px"
-      onClick={() => onSelectItem(index)}
+      onClick={() => !disableFeeSelection && onSelectItem(index)}
```
I haven't tested this myself in the UI, but why would you even see the selector if selecting the fee is disabled?

fbwoolf commented on src/app/components/fees-row/fees-row.tsx:
```diff
@@ -46,7 +49,12 @@ export function FeesRow(props: FeeRowProps): React.JSX.Element {
+        convertAmountToBaseUnit(new BigNumber(Number(props.defaultFeeValue)), STX_DECIMALS)
+      );
+      feeTypeHelper.setValue(FeeTypes[FeeTypes.Custom]);
+    } else if (hasFeeEstimates && !feeField.value && !isCustom) {
```
I think this would read better as it's own if statement rather than useing `else if`.

fbwoolf commented on src/app/features/stacks-transaction-request/stacks-transaction-signer.tsx:
```diff
@@ -0,0 +1,129 @@
+  stacksTransaction: StacksTransaction;
+  disableFeeSelection?: boolean;
+  disableNonceSelection?: boolean;
+
```
Remove empty line here.

fbwoolf commented on src/app/pages/rpc-sign-stacks-transaction/rpc-sign-stacks-transaction.tsx:
```diff
@@ -0,0 +1,186 @@
+  return transactionRequest;
+};
+
+function useRpcSignStacksTransactionParams() {
```
These hooks should be pulled out into their own files.

fbwoolf commented on src/app/pages/rpc-sign-stacks-transaction/rpc-sign-stacks-transaction.tsx:
```diff
@@ -0,0 +1,186 @@
+  const [searchParams] = useSearchParams();
+  const { origin, tabId } = useDefaultRequestParams();
+  const requestId = searchParams.get('requestId');
+  const transactionHex = searchParams.get('transactionHex');
```
Seems like `txHex` would be ok here?

fess-v commented:
Conducted some more testing, fixed a few small things in the last commit FYI @fbwoolf @kyranjamie

fess-v commented:
conflicts fixed too @kyranjamie

kyranjamie commented on src/app/features/stacks-transaction-request/stacks-transaction-signer.tsx:
```diff
@@ -0,0 +1,128 @@
+  onSignStacksTransaction(fee: number, nonce: number): void;
+}
+
+export function StacksTransactionSigner(props: StacksTransactionSignerProps) {
```
Can you destructure props like we do elsewhere?

fess-v commented:
@kyranjamie done

fess-v commented:
@kyranjamie updated the docs in the comment above with the latest parameters we confirmed

janniks commented on src/shared/rpc/methods/sign-stacks-transaction.ts:
```diff
@@ -0,0 +1,36 @@
+import { formatValidationErrors, getRpcParamErrors, validateRpcParams } from './validation.utils';
+
+const rpcSignStacksTransactionParamsSchema = yup.object().shape({
+  stxAddress: yup.string().required(),
```
Should this be optional to allow using current account or selecting afterwards?

janniks commented:
Would it make sense to add namespaces to the rpc for this?
`await window.stacks.request('signStacksTransaction', requestParams);` has "stacks" twice. What are folks thinking RE `stx_signTransaction` (if the provider is NOT on window.stacks) and simply `signTransaction` if we're already on window.stacks (or will window.stacks use other networks as well? one provider for all, or multiple providers?)

fess-v commented:
Hey @janniks @kyranjamie !
Just made the `stxAddress` optional, and also fixed two minor issues that we found while preparing our Stacks Multisig for the release
Removed the context file and set a request parameter instead, making it available in all components instead of only those that are wrapped by the context, It caused a few issues
Also removed balance errors for multisig transactions since owners can have empty wallets but still make transactions through a multisig
Regarding the comment from @janniks about RPC calls
Now the Stacks transaction signature works with this interface:
`const result = await window.btc.request('signStacksTransaction', requestParams);`
We can change it to this:
`const result = await window.stacks.request('signTransaction', requestParams);`
For it, we'll need to modify this file and make handlers separated for btc and stacks, so if you guys think we need to do it I'll make this change as well
![image](https://github.com/leather-wallet/extension/assets/26626714/1efeccb6-2bc6-465d-a42f-31d00d0e8ec0)

fess-v commented:
@kyranjamie @janniks Hey guys! added basic autotests, and made it similar to existing ones, let me know if it looks fine
For the full coverage a lot of things should be added like fee selection, nonce selection, sponsored transactions, etc, but Idk if it's required now, was looking at other autotests and decided to send the current version

fess-v commented:
@kyranjamie just fixed conflicts, took a while

kyranjamie commented:
Made a duplicate PR here with tidied commits

fess-v commented:
thanks a lot @kyranjamie
